executor/write.go

```diff
@@ -154,28 +154,30 @@ func updateRecord(ctx sessionctx.Context, h int64, oldData, newData []types.Datu
 		if err != nil {
 			return false, handleChanged, newHandle, 0, errors.Trace(err)
 		}
-		newHandle, err = t.AddRecord(ctx, newData, skipHandleCheck)
+		// the `affectedRows` is increased when adding new record.
+		newHandle, err = t.AddRecord(ctx, newData, skipHandleCheck, true)
+		if err != nil {
+			return false, handleChanged, newHandle, 0, errors.Trace(err)
+		}
+		if onDup {
+			sc.AddAffectedRows(1)
+		}
 	} else {
 		// Update record to new value and update index.
 		err = t.UpdateRecord(ctx, h, oldData, newData, modified)
-	}
-	if err != nil {
-		return false, handleChanged, newHandle, 0, errors.Trace(err)
-	}
-
-	tid := t.Meta().ID
-	ctx.StmtAddDirtyTableOP(DirtyTableDeleteRow, tid, h, nil)
```

Comment:
Why dirty table operation is changed, it should be irrelevant

Reply:
the row add to dirty table have move to `AddRecord` and `UpdateRecord` and `RemoveRecord`.
Because `AddRecord` will backfill default value to row for write-only column. If dirty table doesn't have the value of write-only column, update will panic in transaction. See https://github.com/pingcap/tidb/pull/8792 description Problem 1.